Guidelines for writing bash

raoulmillais opened the pull request:
What I am suggesting:

- [x] Gets out of the way;
- [x] Is not restricting;
- [x] Is safe, production ready, and project members have experience with it;
- [x] Promotes good coding practices.

- Hat tip to @sublimino for some of this advice

tomgco commented on bash-code-style.md:
```diff
@@ -0,0 +1,161 @@
+process bash is a good choice.  It can also be helpful for prototyping quickly
+before reaching for a more powerful language.  As a rule of thumb bash scripts
+should never grow larger than a few hundred lines.  You probably want a better
+language at that point.
```
Maybe a bit too opinionated, maybe  "You probably want a better language at that point." could be rephrased to: "At this stage you might want to re-evaluate choosing bash"

raoulmillais replied:
Yep I see your point. It's verges towards flamey for a standards document. :+1:

tomgco commented on bash-code-style.md:
```diff
@@ -0,0 +1,161 @@
+
+## Bash code style
+
+In general follow [google's excellent recommendations](https://google.github.io/styleguide/shell.xml)
```
google's -> Google's

raoulmillais replied:
:+1: Thanks

sublimino commented:
- Include testing framework https://github.com/sstephenson/bats and https://github.com/ztombol/bats-assert
- set script to own directory can be useful for relieve path includes `cd "$(dirname "$0")"`
- I've got an options parser I'll link in
- A standard library would be useful, mine's currently in flux so this is perhaps a spike
- I'm using a yeoman generator for scripts atm too - this auto generates a usage from comments in the file. Can paste in when not on mobile if interested
- couldn't see debug mode mentioned at first skim, can be handy `set -x`
- there's a BASH debugger if things get hairy http://bashdb.sourceforge.net
- osx still ships with v3, could be worth highlighting
- I <3 BASH

raoulmillais commented:
Thanks @tomgco and @sublimino.  I incorporated your suggestions.  @sublimino I incorporated your debugging tips, please do submit a pull request to expand and add any other tips :)
